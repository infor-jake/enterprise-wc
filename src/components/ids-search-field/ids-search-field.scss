@import '../ids-trigger-field/ids-trigger-field';

// Applies to search fields in the app menu, which are
// typically on top of black or dark color background
@mixin ids-app-menu-search-field-shadow() {
  box-shadow: var(--ids-search-field-app-menu-shadow-focus);
}

// Applies to both `dark` and `contrast` theme modes
@mixin ids-search-field-dark-app-menu-colors() {
  &.color-variant-app-menu {
    color: var(--ids-search-field-app-menu-color-text);

    .field-container {
      background-color: var(--ids-search-field-app-menu-color-background);
      color: var(--ids-search-field-app-menu-color-text);
    }

    .ids-input-field {
      color: var(--ids-search-field-app-menu-color-text);
    }

    .btn-clear {
      color: var(--ids-search-field-app-menu-color-icon-default);
    }

    .btn-clear:hover {
      color: var(--ids-search-field-app-menu-color-icon-hover);
    }

    ids-icon::part(search-icon) {
      color: var(--ids-search-field-app-menu-color-text);
    }

    &:focus-within {
      ids-icon::part(search-icon) {
        color: var(--ids-search-field-app-menu-color-icon-hover);
      }
    }

    &:focus-within,
    &.readonly:focus-within {
      .field-container {
        @include ids-app-menu-search-field-shadow();
      }
    }
  }
}

.ids-search-field:focus-within ids-input::part(input) {
  box-shadow: none;
}

:host {
  ids-menu-button {
    transition: border var(--ids-motion-duration) var(--ids-motion-ease);
    border: 0;
    border-inline: 0;
    border-inline-end: 1px solid var(--ids-search-field-color-border-default);
    height: 100%;
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    align-content: center;
    margin: 0;
    margin-inline: 0;
    padding: 0;
    padding-inline: 0;

    & > * {
      display: inline-flex;
      align-self: center;
    }

    &::part(button) {
      border: 0;
      border-radius: 0;
      margin: 0;
      margin-inline: 0;
      padding: 0;
      padding-inline: 0;
      padding-inline: var(--ids-search-field-icon-margin);

      &,
      &:active,
      &:hover {
        background: transparent;
        background-color: transparent;
      }
    }
  }
}

:host(:focus-within),
:host(:focus-visible) {
  ids-menu-button {
    border-color: var(--ids-search-field-color-border-focus);

    &::part(button) {
      color: var(--ids-search-field-button-color-text-hover);
    }
  }
}

:host(:active),
:host(:hover) {
  ids-menu-button {
    border-color: var(--ids-search-field-color-border-hover);
  }
}

.ids-search-field {
  .field-container {
    background-color: var(--ids-search-field-color-background-default);
    border-color: var(--ids-search-field-color-border-default);
    border-radius: var(--ids-search-field-border-radius);
    margin: var(--ids-search-field-margin);
  }

  ids-icon.search-icon {
    margin-inline-start: var(--ids-search-field-icon-margin);
  }

  .fieldset {
    display: inline-flex;
    flex-direction: row;
    width: 100%;
  }

  #category-action-button::part(button) {
    padding: var(--ids-space-40);
    margin-inline-start: var(--ids-search-field-icon-margin);
  }

  // =============================
  // Default Styles for color variants (Applicable to all themes)
  &.color-variant-alternate {
    .field-container {
      border-style: none;
      background-color: var(--ids-search-field-header-color-background);
      color: var(--ids-search-field-header-color-text);
   }

    ids-icon {
      color: var(--ids-search-field-header-color-icon-default);
    }
  }

  &.color-variant-card {
    .field-container {
      background: var(--ids-search-field-card-color-background);
      color: var(--ids-search-field-card-color-text);
    }

    .btn-clear:hover {
      background-color: transparent;
    }
  }

  &.color-variant-app-menu {
    color: var(--ids-search-field-app-menu-color-text);

    .field-container {
      border-style: none;
      border-radius: var(--ids-search-field-border-radius);
      background-color: var(--ids-search-field-app-menu-color-background);
      color: var(--ids-search-field-app-menu-color-text);
    }

    .ids-input-field {
       color: var(--ids-search-field-app-menu-color-text);
    }

    .btn-clear {
      color: var(--ids-search-field-app-menu-color-icon-default);
    }

    .btn-clear:hover {
      color: var(--ids-search-field-app-menu-color-icon-hover);
    }

    ids-icon::part(search-icon) {
      color: var(--ids-search-field-app-menu-color-text);
    }

    &:focus-within {
      // stylelint-disable-next-line
      ids-icon::part(search-icon) {
        color: var(--ids-search-field-app-menu-color-icon-hover);
      }
    }

    &:focus-within,
    &.readonly:focus-within {
      // stylelint-disable-next-line
      .field-container {
        @include ids-app-menu-search-field-shadow();
      }
    }
  }

  // Ignore the fieldset element on Module Nav searchfields
  &.color-variant-module-nav {
    .fieldset {
      display: contents;
    }

    .field-container {
      background-color: var(--ids-module-nav-search-input-bg-color);
      color: var(--ids-module-nav-item-hover-text-color);
    }

    .ids-input-field {
      padding-inline-start: var(--ids-module-nav-item-expanded-icon-text-spacing-size);
    }

    &.full {
      .field-container {
        width: var(--ids-module-nav-search-input-width);
      }
    }

    &:focus-within {
      .field-container {
        background-color: var(--ids-module-nav-search-input-focused-bg-color);
        border-color: var(--ids-module-nav-item-selected-text-color);
      }
    }
  }
}
