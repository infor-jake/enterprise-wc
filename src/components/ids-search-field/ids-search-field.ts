import { attributes } from '../../core/ids-attributes';
import { customElement, scss } from '../../core/ids-decorators';
import { stripHTML } from '../../utils/ids-xss-utils/ids-xss-utils';

import IdsTriggerField from '../ids-trigger-field/ids-trigger-field';

import '../ids-trigger-field/ids-trigger-button';
import '../ids-input/ids-input';
import '../ids-icon/ids-icon';
import '../ids-menu-button/ids-menu-button';
import '../ids-popup-menu/ids-popup-menu';

import styles from './ids-search-field.scss';

const DEFAULT_LABEL = 'Search';
const DEFAULT_PLACEHOLDER = 'Type to search';

/**
 * IDS Search Field Component
 * @type {IdsSearchField}
 * @inherits IdsElement
 */
@customElement('ids-search-field')
@scss(styles)
export default class IdsSearchField extends IdsTriggerField {
  constructor() {
    super();
  }

  isFormComponent = true;

  /**
   * Inherited from `IdsColorVariantMixin`
   * @returns {Array<string>} List of available color variants for this component
   */
  colorVariants = ['alternate', 'app-menu', 'card'];

  /**
   * Inherited from `IdsColorVariantMixin`. If the Color Variant on Search Fields are changed,
   * switch trigger buttons to the "alternate" style instead of an `app-menu` style.
   * @param {string} variantName the new color variant being applied to the Search Field
   */
  onColorVariantRefresh(variantName?: string) {
    let btnVariantName = variantName || '';
    if (variantName === 'app-menu') {
      btnVariantName = 'alternate';
    }
    const adjustBtnVariant = (btn: any) => {
      btn.setAttribute(attributes.COLOR_VARIANT, btnVariantName);
    };

    this.buttons.forEach(adjustBtnVariant);
    if (this.fieldContainer) {
      [...this.fieldContainer.querySelectorAll('ids-trigger-button')].forEach(adjustBtnVariant);
    }
  }

  static get attributes() {
    return [
      ...super.attributes,
      attributes.CATEGORY,
    ];
  }

  connectedCallback() {
    super.connectedCallback();
    this.#attachEventHandlers();
    this.#attachKeyboardListener();

    if (!this.placeholder) {
      this.placeholder = DEFAULT_PLACEHOLDER;
    }
    if (!this.label) {
      this.label = DEFAULT_LABEL;
    }
  }

  template(): string {
    this.templateHostAttributes();
    const {
      ariaLabel,
      containerClass,
      inputClass,
      inputState,
      labelHtml,
      placeholder,
      type,
      value
    } = this.templateVariables();

    const searchIcon = `<ids-icon class="ids-icon search-icon starting-icon" size="medium" icon="search"></ids-icon>`;

    return `<div id="ids-search-field" class="ids-search-field ids-trigger-field ${containerClass}" part="container">
      ${labelHtml}
      <div class="fieldset">
        <div class="field-container" part="field-container">
          ${this.categories.length ? '' : searchIcon}
          <slot name="trigger-start"></slot>
          ${this.templateCategoriesMenu()}
          <input
            part="input"
            id="${this.id}-input"
            ${type}${inputClass}${placeholder}${inputState}
            ${ariaLabel}
            ${value}
            ></input>
          <slot name="trigger-end"></slot>
        </div>
        ${this.templateCategoriesButton()}
      </div>
      ${this.autocomplete ? `
        <ids-popup
          type="menu"
          align="bottom, left"
          align-target="#${this.id}-input"
          part="popup"
        >
          <ids-list-box slot="content"></ids-list-box>
        </ids-popup>` : ''}
    </div>`;
  }

  templateCategoriesButton(): string {
    if (!this.action) return ``;

    return `
      <ids-button id="category-action-button" appearance="secondary">
        <span>${this.action}</span>
      </ids-button>
    `;
  }

  templateCategoriesMenu(): string {
    if (!this.categories.length) return ``;

    const menuButtonText = this.category
      ? `<span>${this.category}</span>`
      : `<span class="audible">Select Search Category</span>`;

    return `
      <ids-menu-button id="menu-button" appearance="tertiary" menu="category-menu" icon="search" dropdown-icon>
        ${menuButtonText}
      </ids-menu-button>
      <ids-popup-menu id="category-menu" target="menu-button" trigger-type="click" align="bottom, right">
        <ids-menu-group select="${this.multiple ? 'multiple' : 'single'}" keep-open="true">
          ${this.categories.map((category, idx) => `<ids-menu-item value="${idx}">${category}</ids-menu-item>`).join('')}
        </ids-menu-group>
      </ids-popup-menu>
    `;
  }

  #updateCategoriesMenu() {
    const category = this.category;
    const menuButton = this.#categoriesMenuButton;
    if (!category || !menuButton) return;

    const selectedCategories = this.selectedCategories;

    if (selectedCategories.length === 1) {
      menuButton.text = selectedCategories[0];
    } else if (selectedCategories.length > 1) {
      menuButton.text = `${selectedCategories.length} ${this.localeAPI?.translate?.('Selected')}`;
    } else {
      menuButton.text = category;
    }

    const popup = this.#categoriesPopup?.popup;
    const popupWidth = parseInt(popup?.style?.getProperty?.('min-width') ?? 155);
    const popupMinWidth = this.#categoriesMenuButton?.getBoundingClientRect().width ?? popupWidth;
    popup?.style?.setProperty?.('min-width', `${Math.max(popupMinWidth, popupWidth)}px`);

    if (!this.multiple) {
      this.#categoriesPopup?.hide?.();
    }
  }

  get #categoriesPopup(): any {
    return this.shadowRoot?.querySelector('ids-popup-menu');
  }

  get #categoriesMenuButton(): any {
    return this.shadowRoot?.querySelector('ids-menu-button');
  }

  get #categoriesActionButton(): any {
    return this.shadowRoot?.querySelector('ids-button#category-action-button');
  }

  #categories: string[] = [];

  /**
   * @returns {string[]} returns an array of all the categories that appear in the category dropdown
   */
  get categories(): string[] { return this.#categories; }

  /**
   * @param {string[]} value = sets a list of categories that should appear in the category dropdown
   */
  set categories(value: string[]) {
    this.#categories = value;
    this.#rerender();
  }

  /**
   * @param {string} value - sets the initial text that appears in the categories dropdown menu
   */
  set category(value: string) {
    if (value) {
      this.setAttribute(attributes.CATEGORY, value);
    } else {
      this.removeAttribute(attributes.CATEGORY);
    }
    this.#rerender();
  }

  /**
   * @returns {string} - gets the initial text that appears in the categories dropdown menu
   */
  get category(): string {
    return this.getAttribute(attributes.CATEGORY) ?? '';
  }

  /**
   * @returns {string[]} returns an array of the currently selected categories
   */
  get selectedCategories(): string[] {
    const categories = this.categories;
    const selectedValues = this.#categoriesPopup?.getSelectedValues?.() ?? [];
    const selectedCategories = selectedValues.map((categoryKey: number) => categories[categoryKey]);
    return selectedCategories;
  }

  /**
   * @param {string} value - sets title on the search field action-button
   */
  set action(value: string) {
    if (value) {
      this.setAttribute(attributes.ACTION, value);
    } else {
      this.removeAttribute(attributes.ACTION);
    }
    this.#rerender();
  }

  /**
   * @returns {string} - gets the title that appears on the search field action-button
   */
  get action(): string {
    return this.getAttribute(attributes.ACTION) ?? '';
  }

  /**
   * @param {boolean} value - allows multiple categories to be selected
   */
  set multiple(value: boolean) {
    if (value) {
      this.setAttribute(attributes.MULTIPLE, '');
    } else {
      this.removeAttribute(attributes.MULTIPLE);
    }
  }

  /**
   * @returns {boolean} determines whether multiple categories to be selected or not
   */
  get multiple(): boolean {
    return this.hasAttribute(attributes.MULTIPLE);
  }

  /**
   * Programmatically sets the search field's value and performs an optional search function
   * @param {any} val the incoming value to search for
   * @returns {Array<any>} containing search results, if applicable
   */
  async search(val: any): Promise<any> {
    let ret: any = [];
    const safeVal: any = stripHTML(val);
    if (this.#previousSearchValue !== safeVal) {
      if (this.input) this.input.value = safeVal;
      this.#previousSearchValue = safeVal;

      if (typeof this.onSearch === 'function') {
        ret = await this.onSearch(safeVal);
      }
    }
    return ret;
  }

  /**
   * @param {string} val the new value to set
   */
  set value(val: string) {
    super.value = val;

    const newValue = super.value;
    this.#previousSearchValue = newValue;

    if (typeof this.onSearch === 'function') {
      this.onSearch(newValue);
    }
  }

  get value(): string {
    return super.value;
  }

  /**
   * Define this method to carry out search functionality
   * (override this method when implementing an IdsSearchField).
   * @param {any} [val] the value to be searched for
   * @returns {Array<any>} containing matching search results
   */
  onSearch(val: any = undefined): any {
    return [`${stripHTML(val)}`];
  }

  /**
   * @param {any} value
   */
  #previousSearchValue: any = null;

  /**
   * Adds Search Field specific event handlers
   */
  #attachEventHandlers() {
    const handleSearchEvent = (e: any) => {
      this.search(e.target?.value);
    };

    this.offEvent('change', this.input);
    this.onEvent('change', this.input, handleSearchEvent);

    this.offEvent('input', this.input);
    this.onEvent('input', this.input, handleSearchEvent);

    this.offEvent('click', this.#categoriesActionButton);
    this.onEvent('click', this.#categoriesActionButton, () => {
      this.search(this.value);
      this.#triggerCategoriesEvent('search');
    });

    this.offEvent('selected', this.#categoriesPopup);
    this.onEvent('selected', this.#categoriesPopup, () => {
      this.#updateCategoriesMenu();
      this.#triggerCategoriesEvent('selected');
    });

    this.offEvent('deselected', this.#categoriesPopup);
    this.onEvent('deselected', this.#categoriesPopup, () => {
      this.#updateCategoriesMenu();
      this.#triggerCategoriesEvent('deselected');
    });
  }

  /**
   * Listen for enter key to perform search function
   */
  #attachKeyboardListener() {
    this.offEvent('keydown', this.input);
    this.onEvent('keydown', this.input, (event: any) => {
      const shouldSearchOnReturn = !event?.path?.length || !event.path[0].classList || !event.path[0].classList.contains('ids-icon-button');
      if (['Enter'].indexOf(event.code) > -1 && shouldSearchOnReturn) {
        event.preventDefault();
      }

      switch (event.key) {
        case 'Enter':
          if (shouldSearchOnReturn) {
            this.onSearch(this.input?.value);
            this.#triggerCategoriesEvent('search');
          }
          break;
        default:
          break;
      }
    });
  }

  /**
   * Helper to trigger CustomEvent for various event-types
   * @param {'input' | 'search' | 'deselected' | 'selected'} eventType - the type of event to trigger
   */
  #triggerCategoriesEvent(eventType: 'input' | 'search' | 'deselected' | 'selected') {
    this.triggerEvent(eventType, this, {
      detail: {
        elem: this,
        categories: this.categories,
        categoriesSelected: this.selectedCategories,
        value: this.value,
      },
      bubbles: true,
      cancelable: false,
      composed: false
    });
  }

  #rerender() {
    if (this.container) {
      this.container.innerHTML = this.template();
      this.connectedCallback();
    }
  }
}
