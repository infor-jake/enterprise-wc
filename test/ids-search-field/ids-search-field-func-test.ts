/**
 * @jest-environment jsdom
 */
import createFromTemplate from '../helpers/create-from-template';
import '../helpers/resize-observer-mock';
import '../../src/components/ids-search-field/ids-search-field';

const HTMLSnippets = {
  VANILLA_SEARCH_FIELD: (
    `<ids-search-field></ids-search-field>`
  ),
  DISABLED_SEARCH_FIELD: (
    `<ids-search-field disabled label="Pokemon" value="Snorlax"></ids-search-field>`
  ),
  READONLY_SEARCH_FIELD: (
    `<ids-search-field readonly label="Pokemon" value="Lapras"></ids-search-field>`
  ),
  CATEGORY_FULL: (
    `<ids-search-field label="Categories" category="File Types" value="Search term" action="Go" multiple></ids-search-field>`
  ),
  CATEGORY: (
    `<ids-search-field label="Categories" value=""></ids-search-field>`
  ),
  CATEGORY_SHORT: (
    `<ids-search-field label="Categories Short"></ids-search-field>`
  ),
  CATEGORY_MULTIPLE: (
    `<ids-search-field label="Categories (multiple)" multiple></ids-search-field>`
  ),
  CATEGORY_BUTTON: (
    `<ids-search-field label="Categories + Button" value="" action="Go"></ids-search-field>`
  ),
  CATEGORY_SEARCH_TERM: (
    `<ids-search-field label="Categories + Button" value="keyword here" action="Go"></ids-search-field>`
  ),
};

const CATEGORIES = ['Documents', 'Images', 'Audio', 'Video'];

describe('IdsSearchField Component', () => {
  let s: any;

  const createEvent = (type: any, attributes = {}) => {
    const event = new Event(type);
    Object.assign(event, attributes);
    return event;
  };

  const createKeyboardEvent = (keyName: any) => {
    const event = new KeyboardEvent('keydown', { key: keyName });
    return event;
  };

  beforeEach(async () => {
    s = await createFromTemplate(s, HTMLSnippets.VANILLA_SEARCH_FIELD);
  });

  afterEach(async () => {
    document.body.innerHTML = '';
    s?.remove();
  });

  it('renders with no errors', async () => {
    s = await createFromTemplate(s, HTMLSnippets.VANILLA_SEARCH_FIELD);

    const errors = jest.spyOn(global.console, 'error');
    expect(document.querySelectorAll('ids-search-field').length).toEqual(1);

    s.remove();

    s = await createFromTemplate(s, HTMLSnippets.VANILLA_SEARCH_FIELD);
    document.body.appendChild(s);

    expect(errors).not.toHaveBeenCalled();
  });

  it('renders correctly', async () => {
    const elem = await createFromTemplate(s, HTMLSnippets.VANILLA_SEARCH_FIELD);
    elem.label = 'Test Search Field Label';
    elem.value = 'Eevee';
    elem.placeholder = 'Please type a Pokemon';
    document.body.appendChild(elem);
    elem.template();
    expect(elem.outerHTML).toMatchSnapshot();
  });

  it('sets the label correctly', () => {
    expect(s.label).toBe('Search');

    const customLabels = ['Grocery Items', 'Dog Breeds', 'Book Titles'];

    customLabels.forEach((x) => {
      s.label = x;
      expect(s.label).toBe(x);
    });
  });

  it('sets the value correctly', () => {
    expect(s.value).toBe('');

    const customValues = ['Apple', 'Labradoodle', 'Harry Potter'];

    customValues.forEach((x) => {
      s.value = x;
      expect(s.value).toBe(x);
    });
  });

  it('sets the placeholder correctly', () => {
    expect(s.placeholder).toBe('Type to search');

    const customPlaceholders = ['Enter a vegetable name', 'Try typing "poodle"', 'Looking for a book?'];

    customPlaceholders.forEach((x) => {
      s.placeholder = x;
      expect(s.placeholder).toBe(x);
    });
  });

  it('inits readonly and disabled state correctly', async () => {
    s = await createFromTemplate(s, HTMLSnippets.DISABLED_SEARCH_FIELD);
    s = await createFromTemplate(s, HTMLSnippets.READONLY_SEARCH_FIELD);
  });

  it('sets the disabled state correctly', () => {
    expect(s.disabled).toBeFalsy();

    s.disabled = '';
    expect(s.disabled).toBeTruthy();

    s.disabled = 'false';
    expect(s.disabled).toBeFalsy();
  });

  it('sets the readonly correctly', () => {
    expect(s.readonly).toBeFalsy();

    s.readonly = '';
    expect(s.readonly).toBeTruthy();

    s.readonly = 'false';
    expect(s.readonly).toBeFalsy();
  });

  it('listens for input events', () => {
    s.input.dispatchEvent(
      createEvent('input')
    );
  });

  it('listens for keydown events', () => {
    s.input.dispatchEvent(
      createKeyboardEvent('Enter')
    );

    s.input.dispatchEvent(
      createKeyboardEvent('Space')
    );
  });

  it('fires a user-defined `onSearch` method when the value changes', async () => {
    const data = ['one', 'two', 'three', 'four', 'five'];
    s.onSearch = (val: any) => data.filter((item) => item.includes(val));

    let results = await s.search('five');
    expect(results.length).toBe(1);

    results = await s.search('f');
    expect(results.length).toBe(2);
  });

  it('readonly/disabled should be clearable with clearable-forced setting', async () => {
    s = await createFromTemplate(s, HTMLSnippets.DISABLED_SEARCH_FIELD);
    s.clearable = true;
    expect(s.container.querySelector('.btn-clear')).toBeNull();

    s.clearable = false;
    s.clearableForced = true;
    expect(s.container.querySelector('.btn-clear')).not.toBeNull();

    s = await createFromTemplate(s, HTMLSnippets.READONLY_SEARCH_FIELD);
    s.clearable = true;
    expect(s.container.querySelector('.btn-clear')).toBeNull();
    s.clearableForced = true;
    expect(s.container.querySelector('.btn-clear')).not.toBeNull();
  });

  it.skip('renders categories full', async () => {
    s = await createFromTemplate(s, HTMLSnippets.CATEGORY_FULL);

    // s.categories = ['Documents', 'Images', 'Audio', 'Video'];
    s.categories = CATEGORIES;
    expect(s.container.outerHTML).toMatchSnapshot();
  });

  it('shows action button when action prop set', async () => {
    s = await createFromTemplate(s, HTMLSnippets.CATEGORY_BUTTON);
    s.categories = CATEGORIES;
    expect(s.container.querySelector('#category-action-button')).not.toBeNull();
  });

  it('shows full category dropdown when categories set', async () => {
    s = await createFromTemplate(s, HTMLSnippets.CATEGORY_FULL);
    s.categories = CATEGORIES;

    expect(s.container.querySelector('ids-popup-menu')?.textContent).toMatch(CATEGORIES.join(''));
    expect(s.container.querySelector('ids-menu-button').textContent).toMatch('File Types');
  });

  it('shows short category dropdown when categories set', async () => {
    s = await createFromTemplate(s, HTMLSnippets.CATEGORY_SHORT);
    s.categories = CATEGORIES;

    expect(s.container.querySelector('ids-popup-menu')?.textContent).toMatch(CATEGORIES.join(''));
    expect(s.container.querySelector('ids-menu-button').textContent).toMatch('Select Search Category');
  });

  it('triggers "change" event when input-value changed', async () => {
    s = await createFromTemplate(s, HTMLSnippets.CATEGORY_SEARCH_TERM);
    s.categories = CATEGORIES;

    const changeEventListener = jest.fn((evt) => {
      expect(evt.detail.value).toBe('new keyword here');
    });

    s.addEventListener('change', changeEventListener);

    s.value = 'new keyword here';
    expect(changeEventListener).toBeCalledTimes(2);
  });

  it('triggers "search" event when action-button clicked', async () => {
    s = await createFromTemplate(s, HTMLSnippets.CATEGORY_SEARCH_TERM);
    s.categories = CATEGORIES;

    const searchEventListener = jest.fn((evt) => {
      expect(evt.detail.categories).toBe(CATEGORIES);
      expect(evt.detail.categoriesSelected).toMatchObject([]);
      expect(evt.detail.value).toBe('keyword here');
    });

    s.addEventListener('search', searchEventListener);
    await s.container.querySelector('#category-action-button').click();
    expect(searchEventListener).toBeCalledTimes(1);
  });

  it('triggers "search" event when Enter key pressed', async () => {
    s = await createFromTemplate(s, HTMLSnippets.CATEGORY_SEARCH_TERM);
    s.categories = CATEGORIES;

    const searchEventListener = jest.fn((evt) => {
      expect(evt.detail.categories).toBe(CATEGORIES);
      expect(evt.detail.categoriesSelected).toMatchObject([]);
      expect(evt.detail.value).toBe('keyword here');
    });

    s.addEventListener('search', searchEventListener);

    s.input.dispatchEvent(
      createKeyboardEvent('Enter')
    );

    expect(searchEventListener).toBeCalledTimes(1);
  });

  it('triggers "selected/deselcted" event when category-menu clicked', async () => {
    s = await createFromTemplate(s, HTMLSnippets.CATEGORY_FULL);
    s.categories = CATEGORIES;

    const selectedEventListener = jest.fn((evt) => {
      expect(evt.detail.categories).toBe(CATEGORIES);
      expect(evt.detail.categoriesSelected).toMatchObject(['Images']);
      expect(evt.detail.value).toBe('Search term');
    });

    const deselectedEventListener = jest.fn((evt) => {
      expect(evt.detail.categories).toBe(CATEGORIES);
      expect(evt.detail.categoriesSelected).toMatchObject([]);
      expect(evt.detail.value).toBe('Search term');
    });

    s.addEventListener('selected', selectedEventListener);
    s.addEventListener('deselected', deselectedEventListener);

    await s.container.querySelector('ids-popup-menu').items[1].click();
    expect(selectedEventListener).toBeCalledTimes(1);

    await s.container.querySelector('ids-popup-menu').items[1].click();
    expect(deselectedEventListener).toBeCalledTimes(1);
  });

  it('updates menu-button text to say # selected', async () => {
    s = await createFromTemplate(s, HTMLSnippets.CATEGORY_FULL);
    s.categories = CATEGORIES;

    await s.container.querySelector('ids-popup-menu').items[2].click();
    expect(s.container.querySelector('ids-menu-button').textContent).toMatch('Audio');

    await s.container.querySelector('ids-popup-menu').items[3].click();
    expect(s.container.querySelector('ids-menu-button').textContent).toMatch('2 Selected');

    await s.container.querySelector('ids-popup-menu').items[2].click();
    expect(s.container.querySelector('ids-menu-button').textContent).toMatch('Video');
  });
});
