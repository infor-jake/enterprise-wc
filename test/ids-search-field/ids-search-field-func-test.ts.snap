// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsSearchField Component renders categories full 1`] = `
"<div id="ids-search-field" class="ids-search-field ids-trigger-field ids-input md field-height-md" part="container"><div id="ids-search-field" class="ids-search-field ids-trigger-field ids-input md" part="container">
      <label class="ids-label-text" for="ids-input-20-input" part="label">
      <ids-text part="label" label="true" color-unset="">
        Categories
      </ids-text>
    </label>
      <div class="fieldset">
        <div class="field-container" part="field-container">
          
          <slot name="trigger-start"></slot>
          
      <ids-menu-button id="menu-button" appearance="tertiary" menu="category-menu" icon="search" dropdown-icon="dropdown"><ids-icon icon="search" class="ids-icon"></ids-icon>
        <span>File Types</span>
      </ids-menu-button>
      <ids-popup-menu id="category-menu" target="menu-button" trigger-type="click" align="bottom, right" role="none" hidden="">
        <ids-menu-group select="multiple" keep-open="true" role="group" aria-label="Menu group containing 4 item(s)">
          <ids-menu-item value="0">Documents</ids-menu-item><ids-menu-item value="1">Images</ids-menu-item><ids-menu-item value="2">Audio</ids-menu-item><ids-menu-item value="3">Video</ids-menu-item>
        </ids-menu-group>
      </ids-popup-menu>
    
          <input part="input" id="ids-input-20-input" type="text" class="ids-input-field start readonly-background" placeholder="Type to search" value="Search term">
          <slot name="trigger-end"></slot>
        </div>
        
      <ids-button id="category-action-button" appearance="secondary">
        <span>Go</span>
      </ids-button>
    
      </div>
      
    </div></div>"
`;

exports[`IdsSearchField Component renders correctly 1`] = `"<ids-search-field placeholder="Please type a Pokemon" label="Test Search Field Label" value="Eevee"></ids-search-field>"`;
